Implement performance-counter cmdlets for Windows CoreCLR

jeffbi opened the pull request:
This is a PR to provide code for review

Commit:
Add a minishell clixml tests

Commit:
Improve pipeline for native commands

- Start native process in Prepare() instead of Complete()
  in NativeCommandProcessor.
- Remove unneeded input, output, error threads. Replaced by
  simpler primitives.

Commit:
Add error stream and information stream tests

for native pipes

Francisco-Gamino committed:
Fixing update/save-help on Linux.

Commit:
Make native pipeline consume available output in the process block

Commit:
Add motivation example for the native pipeline changes

Commit:
Fix flushing mechanics for input writer in native pipe

Commit:
Use StreamWriter.Dispose() instead of Close()

Francisco-Gamino committed:
Adding tests for update/save-help, and fixing existing help tests for
Linux.

brycem committed:
Release versions of WindowsServerCore and Nano DockerFiles

Embedded POWERSHELL_* ARGs will need to be updated when new releases
become available

brycem committed:
Updating to use Alpha-11 URLs\SHA

Updated URLs to point at Alpha 11 URLs, reduced number of intermediate
stages, simplified logging code, and added a new %PSCORE% environment
variable to make it easier to launch from other shells

Commit:
Add a test for broken linux pipe in native pipeline

Commit:
Add a code to handle finished process correctly on Unix

Commit:
Clean-up unnesesary fields from NativeCommandProcessor

lzybkr committed:
Remove CheckForSevereException and private FailFast

Our private implementation of FailFast was likely introduced
when each individual application needed it's own code to
get Watson reports. The CLR takes care of this for us now,
so we don't need our implementation.

Our method CheckForSevereException was also introduced in
the early days of the CLR - the exceptions it was checking
for aren't actually raised the CLR anymore, they just FailFast.
I removed them as there is a tiny bit of overhead (in code size)
and also in the generated code, e.g. dynamic sites called the method.

Commit:
Implement performance-counter cmdlets for Windows CoreCLR

This is a PR to provide code for review

Commit:
Implement performance-counter cmdlets for Windows CoreCLR

This is a PR to provide code for review

msftclas commented:
Hi __@jeffbi__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Jeff Bienstadt (Jetstream Software)). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

mirichmo committed:
Adding alpha.12 to the list of spelling exclusions (#2603)

daxian-dbw committed:
Update project version to 6.0.0 to match our release version

daxian-dbw committed:
Add 'AssemblyVersion' to 'GraphicalHost' and 'LocalAccounts' projects

HemantMahawar committed:
Use the unabbreviated parameter names in Start-PSBootstrap (#2623)

KirkMunro committed:
Added -Top/-Bottom params to Sort-Object for Top/Bottom N sort and many Pester tests (#2518)

* Added -Top/-Bottom params to Sort-Object for Top/Bottom N sorts; added many Pester tests

* updated changelog

* updated CHANGELOG.md

* replaced Get-Service with Get-Alias in Pester tests

* improved performance for -unique -top/-bottom, changed according to PR feedback

* added newlines to end of two files

* Removed whitespace changes

* Missed 2 whitespace changes to undo in last commit

* Updated to try to resolve merge issue

* change heapify return type, cleaned up test code

* unique top/bottom N sorts are now 2-3 times faster

* code cleanup (removed unused vars, simplified logic)

charub committed:
Changes to get security zone when urlmon.dll is not present (#2625)

mirichmo committed:
Fixing alpha.11 name change (#2628)

SteveL-MSFT committed:
Update KNOWNISSUES.md (#2624)

Removed select-xml from the unsupported table based on #2316 as it appear to be working on PSCore on windows

SteveL-MSFT committed:
removed Handles column from Process formatter (#2617)

addressing #1812

brycem committed:
Updating to Alpha 12 and correcting 'reslease' typo

Commit:
Replace spin-lock by a proper BlockingCollection in NativeCommandProcessor

Commit:
Add native pipelines entry to changelog.md

adbertram committed:
Fixed broken link in README (#2643)

iSazonov committed:
Improve a progress pane performance (#2640)

* Improve a progress pane performance

Add timer to update a progress pane every 100 ms

Francisco-Gamino commented on src/Microsoft.PowerShell.Commands.Diagnostics/CommonUtils.cs:
```diff
@@ -6,11 +6,14 @@
 using System.Resources;
 using System.Reflection;
 
+#if CORECLR
```
Could you please add tests for these cmdlets?

brycem committed:
Refactoring Linux Nighly Dockerfiles to in-line Bootstrap.ps1 code

Provides simpler usability by making the dockerfiles self-contained.

megamorf committed:
Fix small typo (#2647)

it's -> its

latkin committed:
Wire up proper ConsoleHostRawUserInterface.LengthInBufferCells for Unix (#2637)

* Wire up proper ConsoleHostRawUserInterface.LengthInBufferCells for Unix

Ref https://github.com/PowerShell/PowerShell/issues/2502
This allows for Unix console host to properly calculate display width for
strings containing escape sequences (e.g. ANSI color), rather than
falling back to naive string.Length.

* Add basic tests for host LengthInBufferCells calculation

nanalakshmanan committed:
Fixing background jobs for Unix and Windows (#1972)

iSazonov committed:
Add Get-Uptime to Microsoft.PowerShell.Utility (C# cmdlet) (#2517)

* Add Get-Uptime (C# code)

1. Add Get-Uptime cmdlet (C# code)
2. Modify psd1 files
3. Add tests

* Remove Pretty parameter after Powershell committee conclusion

1. Remove Pretty parameter
2. Remove tests for Pretty
3. Add Throw if  IsHighResolution is False
4. Fix formatting

* Add test to check throw

1. Add hook that simulate 'System.Diagnostics.Stopwatch.IsHighResolution == false' to test Get-Uptime throw
2. Add test to check throw

* Fixes after review

1. Remove Timespan parameter
2. Remove unneeded tests
3. Fix localization
4. Add string consts
5. Fix throw

* Fix OutputType to add parameter set name information

jianyunt committed:
Registry.LocalMachine throws on non-windows platform.

adityapatwardhan committed:
Add OpenCover PS Module to collect code coverage (#2585)

* Add OpenCover PS Module to collect code coverage

OpenCover PS Module helps is collecting Code Coverage using the OpenCover
toolset. The module helps in comparing two code coverage runs as well.

* Change OpenCover.psd1 to ASCII

* Fix an error in path for OpenCover

Fixed an error on path for OpenCover. Also used ZipFile class instead of
cmdlet as it might not be available on CI system.

* Convert module to be Powershell v4 compliant

Changed implementation from classes to PSObjects and implemented
Expand-ZipArchive.

* Added CodeCoverage as a configuration to project.json files

Added CodeCoverage as the new configuration for all the project.json
files. When Start-PSBuild is executed with configuration as CodeCoverage,
we change the degubType to 'full' as required by OpenCover toolset.

Also made changes to appveyor.psm1 to build a CodeCoverage package on
daily builds and publish it as a zip.

* Addressed code review comments

Changed from Add-Member to use pscustomobject type accelator. Removed
[gc]::collect.

* Added explicit garbage collection

* Addressed code review comments

- Make sure that the build Start-PSPackage gets is not a code coverage
  build
- Add debugType = full for FullCLR
- Remove configurations from PackageManagement files as it is not needed.
- Build CodeCoverage build first in AppVeyor.

* Resolve merge conflict

* Fix indentation

* Fix newline at end of file

* Added command discovery for locating OpenCover.console.exe

lzybkr committed:
Make Out-Null fast (#2629)

People use Out-Null despite much faster alternatives:

    $null = Do-Stuff
    [void] = Do-Stuff
    Do-Stuff > $null

This change makes Out-Null work in roughly the same say as the above.

The optimization is to detect that we're calling the built-in Out-Null
cmdlet when invoking a pipeline from script (the change won't have any
effect in the PowerShell api). If we detect Out-Null, we rewrite the
pipe to look similar to `Do-Stuff > $null`.

Commit:
Re-shuffle minishell tests

Commit:
Prettify NativeCommmandProcessor.cs

- Move ProcessOutputHandler outside of NativeCommmandProcessor
- Remove redundent ConsumeAvailableNativeProcessOutput() blocking call

Commit:
Resurect Xml serialization logic inside ProcessInputWriter

Commit:
Fix bug in native pipe, where we loose End() output

Commit:
Fix recursive Done() problem in steppable pipeline finalization

Commit:
Trying to make 'native command lifecycle' test unflaky

Also move it to the Feature category

daxian-dbw committed:
Update releasing doc to include steps to create Linux/macOS packages and NuGet packages (#2663)

* Update releasing.md with package creation steps for Linux and macOS

* Add documentation for creating NuGet Packages

* Minor fixes

lzybkr committed:
Delay initialize some properties in InformationRecord

An InformationRecord instance is created when calling Write-Host,
but many of the properties aren't read.

It can be expensive to initialize some of those properties, so
the code has been changed to initialize those properties on first access.

lzybkr committed:
Improve coverage for Write-Information output

SteveL-MSFT committed:
Update CONTRIBUTING.md

Fixed reference to 0-backlog

SteveL-MSFT committed:
Update CONTRIBUTING.md

fix link

jianyunt committed:
Merge remote-tracking branch 'upstream/master'

jianyunt committed:
address cr feedback

Commit:
Temporarely pin dotnet obtain script url

Commit:
Remove default value for version from Install-Dotnet, so it's controlled only in Start-PSBootstrap

lzybkr committed:
Fix LengthInBufferCells test for downlevel

When running on systems that do not support a virtual terminal,
the tests was incorrectly assuming escape sequences were removed.

The fix is to expect different results depending on whether or not
the host supports virtual terminal mode.

vors committed:
Fixed reference to 0-backlog in CONTRIBUTING.md (#2681)

daxian-dbw committed:
Fix PackageManagement issue in the latest Alpah12 on Linux (#2665)

* Registry.LocalMachine throws on non-windows platform.

* address cr feedback

mirichmo committed:
Add support to build x86 packages for PowerShell Core

Commit:
Fix WebRequest failure to handle missing ContentType in response header (#2666)

If a response to invoke-webrequest does not set the content type in the
response header, an object not set exception is thrown in
WebResponseHelper.CoreClr.cs.  Updated to set contenttype = null.

mirichmo committed:
Make the MSI installers support Win 7 and above

mirichmo committed:
Add runtime win7-x86

mwrock committed:
Ensure *nix platforms add an initial / rooted drive (#2679)

Commit:
Separate Done() and Dispose() in ProcessInputWriter NativeCommandProcessor.cs

SteveL-MSFT committed:
Update issue-management.md (#2673)

* Update issue-management.md

* Update issue-management.md

fixed link to PR code review

Commit:
Remove over-optimization in NativeCommandProcessor.DequeueProcessOutput

daxian-dbw committed:
Merge pull request #2450 from vors/native-pipe

Refactor 'NativeCommandProcessor' to provide streaming behavior for running native commands in a pipeline.

iSazonov committed:
Write-Information accept objects from the pipeline (#2680)

* Write-Information accept objects from the pipeline

* Add test

* Fix test

Remove unneeded runspace.

* Clear streams before each test case

Streams of the PowerShell instance might have records from other test runs. Clear the streams before each test case.

mirichmo committed:
Add x86 support to the MSI installer

iSazonov committed:
Correct samples in testing-guidelines (#2691)

Replace 'Start-PSPester -Directory' with 'Start-PSPester -Path'

daxian-dbw committed:
Add HelpUri for Get-Uptime (#2688)

* Add HelpUri for Get-Uptime

* Update online help test data to include Get-Uptime

mirichmo committed:
Add x86 support to the remoting install script

alexpilotti committed:
Fixes deprecated syscall issue on macOS 10.12 (#2633) (#2675)

syscall() has been deprecated and is not available on macOS 10.12.

SteveL-MSFT committed:
added type to returned objects
replaced expand-ziparchive with call to Archive module
added defaults to location of opencover.xml, opencover installation, powershell.exe, and location of tests
updated error messages with corrective action
replaced System.Net.WebClient call (not available on coreclr) with invoke-webrequest
scoped change of executionpolicy to process when starting pester to avoid error if system policy is less restrictive
updated version of module

vors committed:
Re-enable Add-Type coreclr tests (#2705)

SteveL-MSFT committed:
addressing code review feedback
- check for git before using
- check for opencover.console in path
- fixed path to code coverage build of PS

SteveL-MSFT committed:
added Expand-ZipArchive back to support psv4 scenario

vors committed:
Fix code errors found by PVS-Studio (#2695)

* Fix errors found by PSV-studio analyzer

http://www.viva64.com/en/b/0447/

* Remove old redundent comment

errorAsts is passed by ref

JamesWTruher committed:
Added tests to close code coverage in CIM area (#2528)

* Added tests to close code coverage in CIM area

Added CimAdapter.Tests for adaptation layer (code cribbed from BVT tests)
Added SessionOption.Tests.ps1 (Code cribbed from BVT tests)
Modified CimClass tests by adding more tests
Added CimSession.Tests for basic tests of New-CimSession cmdlet
Added CimInstance.Tests for basic tests of Get-CimInstance

* Add CI tags to tests

* unify test execution to use try/catch for marking tests as skipped

moved test which retrieves cimclass via method to feature as it is not
a common operation, this test can take some time as well

JamesWTruher committed:
Jameswtruher/dailytravis (#2696)

* Add daily build code for travis

A cron job needs to be created to start the build, and the variable
TRAVIS_DAILY_BUILD must be set to true

* Change environment variable to the one suggested by the travis-ci documentation

TRAVIS_EVENT_TYPE == "cron"

jianyunt committed:
Update build.psm1 to pull PackageManagement assemblies

jianyunt committed:
Use alpha.11 instead of alpha.12 to allow using OneGet to install modules from myget

jianyunt committed:
update project.json

jianyunt committed:
Add PackageManagement acceptance test cases

jianyunt committed:
remove PackageManagement source code

jianyunt committed:
remove packagemanagement test code

iSazonov committed:
Add support W3C Extended Log File Format in Import-Csv (#2482)

* Add support W3C Extended Log File Format

1. Add support W3C Extended Log File Format
2. Refactoring Import-Csv tests

3. Add #Type tests

* Fx test after review

* Fix tests after review

* Remove unneeded test

jianyunt committed:
# This is a combination of 3 commits.
# The first commit's message is:

Changed to PSModuleRestore switch, i.e., by default no PSModule install

# This is the commit message #2:

install PowerShell modules to publish folder as well as one level up

# This is the commit message #3:

removed workaround

iSazonov committed:
Now Enum values definition have space-insensitive syntax (#2690)

Fix to support whitespace after the `=` and before the value expression.

oising committed:
Guard against ReflectionTypeLoadException in InitializeTypeCache (#2549)

SteveL-MSFT committed:
updates to opencover module to make it more user friendly and works on coreclr (#2702)

* added type to returned objects
replaced expand-ziparchive with call to Archive module
added defaults to location of opencover.xml, opencover installation, powershell.exe, and location of tests
updated error messages with corrective action
replaced System.Net.WebClient call (not available on coreclr) with invoke-webrequest
scoped change of executionpolicy to process when starting pester to avoid error if system policy is less restrictive
updated version of module

* addressing code review feedback
- check for git before using
- check for opencover.console in path
- fixed path to code coverage build of PS

* added Expand-ZipArchive back to support psv4 scenario

Commit:
Replace travis.sh by travis.ps1

Fix problem with running Start-PSPester from the wrong directory

Commit:
Remove redundent $buildConfiguration in appveyor.psm1

SteveL-MSFT committed:
Merge remote-tracking branch 'upstream/master'

jianyunt committed:
add psmodulerestore in travis.ps1

jianyunt committed:
Removed PowerShellGet source and replaced with ones from myget

Commit:
Make Restore-PSModule -Name parameter a [string[]]

Commit:
Make Start-PSBootstrap on macOS better

No errors if brew dependencies already present

mirichmo committed:
Add x86 CrossGen support (#2726)

vors committed:
Fix broken travis.ps1 logic (#2739)

daxian-dbw committed:
Fix the indicator for redirecting standard input (#2749)

mirichmo committed:
Fixing powershell.exe lib paths and more thoroughly cleaning the powershell-native directory (#2758)

bmanikm committed:
- Fixed failing PackageManagement related tests in daily-builds (#2759)

- Updated Restore-PSModule function to delete the PSGetModuleInfo.xml so that this file will not be included in the PowerShell release.
- Since PackageManagement module gets installed with PowerShellGet module, removed PackageManagement module name in Restore-PSModule command invocation.

Commit:
Code changes per PR code review, and adding Pester tests

Commit:
Code changes per PR code review, and adding Pester tests

daxian-dbw committed:
Update docs with release alpha.13 (#2765)

* Update docs with release alpha.13
* Additional doc updates
* Update the 'Get PowerShell' table

vors committed:
Remove obsolete entries in FAQ.md (#2755)

iSazonov committed:
Fix Verbs<> class strings  #2716 (#2762)

* Fix Verbs
* Replace `-runspace` with `-Session`

SteveL-MSFT committed:
updated error messages to tell user what to do (#2766)

adityapatwardhan committed:
Moved publishing of code coverage artifacts to test_script phase (#2760)

* Moved publishing of code coverage artifacts to after_test phase

The webhook for build completion is called after the after_test phase
hence we needed to publish code coverage artifacts before that. Moved the
logic for compression and publish the artifacts after running tests.

* Added check for daily build

* Address code review comments.

* Addressed code review comments

Addressed comments about definition of artifacts.
Defined new function for after_test phase.

daxian-dbw committed:
Add back -Configuration 'Release' to make sure AppVeyor CIs produce 'Release' bits (#2773)

adityapatwardhan committed:
Assigned all ArrayList.Add() to $null (#2776)

* Assigned all ArrayList.Add() to $null
ArrayList.Add() returns the array list size after the add. This causes the
value to be thrown on the pipeline. Assigning them to $null as we do not
need to know the size after addition. This causes errors and unnecessary
output on the AppVeyor console.
* Changed the way nuget artifacts are added to arraylist
* Remove unnecessary `$null =` for `AddRange` call because it's a void method.

adityapatwardhan committed:
Added -erroraction silentlycontinue for import-module (#2778)

If the module is not available it throws an error. To make sure OpenCover
works on PSv4 and below, -erroraction silentlycontinue was added.

KirkMunro committed:
Fixed typo (#2782)

You might want to correct this in the notes associated with the release as well (I don't seem to be able to edit that).

lzybkr committed:
Use -split operator to avoid error from new overload (#2796)

adityapatwardhan committed:
Added -publish to codecoverage package generation and fixed package name (#2797)

* Added -publish to codecoverage package generation and fixed package name

Added -Publish to Start-PSBuild used by CodeCoverage configuration so that
.netcore dependencies are also included in the package. This makes the
package usable on systems that do not have .netcore in path.

Fixed the name of the coverage package by making it consistent with other
package names.

* Created a function to generate packagename

Francisco-Gamino commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/Export-Counter.Tests.ps1:
```diff
@@ -0,0 +1,216 @@
+                }
+                $cmd = "$cmdletName $pathParam $formatParam -InputObject `$counterValues $($testCase.Parameters) -ErrorAction Stop"
+
+                if ($testCase.CreateFirst)
```
Could you please add some comments about what CreateFirst means

Francisco-Gamino commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/Export-Counter.Tests.ps1:
```diff
@@ -0,0 +1,216 @@
+                    $pathParam = "-Path $filePath"
+                }
+                $cmd = "$cmdletName $pathParam $formatParam -InputObject `$counterValues $($testCase.Parameters) -ErrorAction Stop"
+
```
Please add a write-host "Command to run: $cmd". This will help debugging without having to run the test case.

Francisco-Gamino commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/Export-Counter.Tests.ps1:
```diff
@@ -0,0 +1,216 @@
+    BeforeAll {
+        $cmdletName = "Export-Counter"
+
+        . "$PSScriptRoot/CounterTestsCommon.ps1"
```
Please add the helper functions in the same file. You can add them before the describe block.

jeffbi replied:
The helper functions are used by multiple test scripts. I have, as we discussed, renamed the file to CounterTestHelperFunctions.ps1, and made additional changes in that file.

Francisco-Gamino commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/CounterTestsCommon.ps1:
```diff
@@ -0,0 +1,182 @@
+    return $counterHelper.TranslateCounterPath($counterName)
+}
+
+function TranslateCounter($counterName)
```
Could you please add a comment about that is the purpose of this function?

Francisco-Gamino commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/CounterTestsCommon.ps1:
```diff
@@ -0,0 +1,182 @@
+}
+"@
+Add-Type -TypeDefinition $helperSource
+
```
Please use single spacing.

jeffbi replied:
Done

Francisco-Gamino commented on src/Microsoft.PowerShell.Commands.Diagnostics/PdhSafeHandle.cs:
```diff
@@ -21,7 +24,9 @@ public override bool IsInvalid
         }
 
 
+#if !CORECLR
```
Could you please add a comment of why we need this on CoreCLR?

Francisco-Gamino commented on src/Microsoft.PowerShell.Commands.Diagnostics/PdhSafeHandle.cs:
```diff
@@ -44,7 +49,9 @@ public override bool IsInvalid
         }
 
 
+#if !CORECLR
```
Same here and below.

daxian-dbw committed:
Revert the timer change introduced in #2640 because it causes race conditions

daxian-dbw committed:
Temporarily disable Update-Help and Save-Help tests in CI builds
Also fix a flaky Start-Sleep test

HemantMahawar committed:
Update issue-management.md (#2823)

* Added description for 2 'committee' tags
* Fixing slashes direction

JamesWTruher committed:
modify Start-PSPester to accept -Quiet to eliminate Pester output (#2795)

* modify Start-PSPester to accept -Quiet to eliminate Pester output

also modify travis.ps1 to include -Quiet for Pester args to reduce
stdout log size which should hopefully allow daily builds to run
without being cancelled due to too much output

* Add error detail information as part of Test-PSPesterResults

Test-PSPesterResults will now also emit information about the errors
found during a test run rather than just the number of failed tests.
Created a new errorlog function so the output of errors will be red.

Commit:
Moving New-Guid to compiled cmdlet (#2752)

lzybkr committed:
Remove unneeded blocks

adityapatwardhan committed:
Added check for empty $nugetArtifacts (#2817)

When the build fails there are no $nugetArtifacts, so AddRange fails

jeffbi commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/Export-Counter.Tests.ps1:
```diff
@@ -0,0 +1,216 @@
+                }
+                finally
+                {
+                    if ($filePath)
```
This test was not meant to determine if a file exists, just whether the $filePath variable contains a path, to prevent invoking Remove-Item with a null or empty path. I can remove the test if you like.

adityapatwardhan committed:
Add ErrorAction Stop so DeviceGuard function (#2825)

A non-terminating error is thrown when the namespace does not exist by
Get-CimInstance. The namespace is not present on OS versions below Windows
10. It is not caught in the try-catch unless ErrorAction Stop is added.

JonathanMalek committed:
Wrapping a call to Registry so that Invoke-WebRequest does not fail on non-text responses (#2811)

* Wrapping a call to Registry with a check for Platform.IsWindows

lzybkr committed:
Remove one more extra Expression.Block call (#2833)

PaulHigin committed:
SSH remoting cmdlet parameter set changes based on RFC 0010 (#2710)

* RFC 0010 SSH remoting cmdlet updates.

* Tweaks to SSH remoting cmdlets

* SSHTransport parameter is now optional instead of mandatory.  Aded fix for CoreCLR entry point when hosted in SSH as a subsystem.

* Added tests

* Code review feedback

* Code review comment changes

* Making SSHTransport switch parameter mandatory because otherwise existing Invoke-Command parameter set

* The new SSH parameter set was causing legacy positional based parameters to no longer work. This fix separates the SSH parameter sets from the existing WinRM parameter sets.

* Fixed region name spelling error.

adityapatwardhan committed:
Disable help content test for Microsoft.Powershell.Archive module (#2835)

The published help content filename is missing '.psm1', but the help still
works. The bug is in PlatyPS and needs to be fixed to re-enable this test.

iSazonov committed:
Implicitly convert value assigned to XML property to string (#2678)

Using LanguagePrimitives.ConvertTo (to emulate what would happen if the property were a regular .Net property typed as string), we now convert the value assigned to an XML property to string.

Commit:
Changes per code review

Commit:
Changes per code review

jeffbi commented on src/Microsoft.PowerShell.Commands.Diagnostics/CoreCLR/Stubs.cs:
```diff
@@ -0,0 +1,281 @@
+using System.ComponentModel;
+using Microsoft.PowerShell.CoreClr.Stubs;
+
+
```
Fixed

jeffbi commented on src/Microsoft.PowerShell.Commands.Diagnostics/CoreCLR/Stubs.cs:
```diff
@@ -0,0 +1,281 @@
+		{
+		}
+
+		public override TypeConverter.StandardValuesCollection GetStandardValues(ITypeDescriptorContext context)
```
This function and its containing class have been removed

jeffbi commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/Export-Counter.Tests.ps1:
```diff
@@ -0,0 +1,216 @@
+            }
+        }
+    }
+
```
Done

jeffbi commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/Export-Counter.Tests.ps1:
```diff
@@ -0,0 +1,216 @@
+            @{
+                Name = "Fails when given invalid path"
+                Path = "c:\DAD288C0-72F8-47D3-8C54-C69481B528DF\counterExport.blg"
+                Parameters = ""
```
Done

adityapatwardhan committed:
Marked flaky test as pending to unblock daily builds (#2838)

dchristian3188 committed:
Updated Split-Path to work with UNC Roots (#2788)

daxian-dbw committed:
Depend on the NuGet package 'libpsrpclient' to enable client-side PSRP on Linux and OSX

mirichmo commented on src/Microsoft.PowerShell.Commands.Diagnostics/ExportCounterCommand.cs:
```diff
@@ -156,15 +156,16 @@ protected override void BeginProcessing()
+            if (osVersion.Major < 6 ||
+                (osVersion.Major == 6 && osVersion.Minor < 1))
             {
                 string msg = _resourceMgr.GetString("ExportCtrWin7Required");
```
PowerShell Core runs on Win 7 and above

mirichmo commented on src/Modules/Windows-Core/Microsoft.PowerShell.Diagnostics/Diagnostics.format.ps1xml:
```diff
@@ -0,0 +1,79 @@
+﻿<Configuration>
+   
+
+  <!-- ################ VIEW DEFINITIONS ################ -->
```
Are these new or were they ported from SD?

jeffbi replied:
These were copied from a different tree location after a comment in the first PR commit, and a subsequent conversation, from Francisco.

daxian-dbw committed:
Add basic test to make sure 'New-PSSession' doesn't crash powershell

mirichmo commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/CounterTestHelperFunctions.ps1:
```diff
@@ -0,0 +1,310 @@
+
+    # the first item exported always seems to have several empty values
+    # when it should not, so we'll start at the second item
+    for ($i = 1; $i -lt $setA.Length; $i++)
```
The comment on line 293 and 294 is a bit concerning. Does that indicate a bug in the code? Will the comparison not work if you start at index zero? Can the problematic export values we skipped from comparison?

Or does this indicate a bug in the test scripts?

jeffbi replied:
I looked a little more deeply into this after putting up the commit. It doesn't appear to be a bug _per se_ in the cmdlet nor in the test script. Experimenting with the Pdh functions in native code, I see the same results in the exported file---for some counters, their entry in the first exported record contains " " rather than a numeric value. It seems to be based on the type of counter being exported.

It might be appropriate to modify the comment now that I know a bit more about what's behind the issue.

jeffbi replied:
I have modified the comment to be somewhat more informative

mirichmo commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/Get-Counter.Tests.ps1:
```diff
@@ -0,0 +1,231 @@
+        # in a known set of values or evan a set with a known minimum
+        # number of items.
+        It "Can process counter set name with wildcards" {
+            $wildcardBase = "roc"
```
These test should then check to make sure that the current language is English and then conditionally execute if it matches. -Skip works well for this technique.

jeffbi replied:
Didn't know about -Skip. I'll look into this and make the appropriate changes.

mirichmo commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/Import-Counter.Tests.ps1:
```diff
@@ -0,0 +1,433 @@
+        $corruptBlgPath = "$outputDirectory/assets/CorruptBlg.blg"
+        $notFoundPath = "$outputDirectory/DAD288C0-72F8-47D3-8C54-C69481B528DF.blg"
+
+        Write-Host "Gathering counter values for export..."
```
Unnecessary print statement. Please remove

jeffbi replied:
I put that in only because the gathering of counter data takes 25 seconds to run. I'll go ahead and remove it.

jeffbi replied:
Removed line

mirichmo commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/Import-Counter.Tests.ps1:
```diff
@@ -0,0 +1,433 @@
+                }
+            }
+            @{
+                # This test should work for English, but other languages are
```
Same -Skip comment here

jeffbi replied:
Fixed

JamesWTruher committed:
Jameswtruher/travisdaily2 (#2842)

* Assign $ProgressPreference = "SilentlyContinue" to reduce output size

Travis is still complaining that the log file is too big, eliminating progress is
another step to reducing output. It's also not needed for our CI environment.

* Removed white space at end of lines

* fix misplacement of progress preference setting

also fix up trailing whitespace

kittholland committed:
Rewrite New-TemporaryFile in C# (#2786)

Literal translation of New-TemporaryFile function to a C# cmdlet,
other than moving logic from Begin to End.

randomvariable committed:
Rebase onto .NET Core 1.1 (#2737)

* Rebase onto .NET Core 1.1

Modify `Build.psm1` and `project.json` files to use .NET Core 1.1.

.NET Core 1.1 ships with an older dotnet/cli than has currently been used,
so we revert to use case-sensitive directory names for dependencies.

.NET Core 1.1 is a pre-requisite for supporting Fedora 24.

* PSReadLine: Bump major version of PSReadline to 6.0.0-*

* Microsoft.PowerShell.Commands.Utility: Bump major version of Microsoft.CodeAnalysis.Csharp

* Microsoft.PowerShell.SDK: Revert attempt to import netcoreapp1.1 instead of dnxcore

* build.psm1: Restore ability to specify version of dotnet-install.sh

adityapatwardhan committed:
Fix OpenCover.console.exe parameter ordering issue. (#2854)

* Fix OpenCover.console.exe parameter ordering issue.

The targetargs parameter needs to be at the end, otherwise everything
after that is considered as part of targetargs.

* Addressed code review comments

* Removed unnecessary file

dlwyatt committed:
Adding ValidateNotNullOrEmpty attribute to parameters listed in #2672 (#2685)

jeffbi commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/CounterTestHelperFunctions.ps1:
```diff
@@ -0,0 +1,310 @@
+        $objectName = $null
+        $counterName = $null
+        $id = Get-PerformanceCounterID $parts.ObjectName
+        if ($id -ne $null)
```
I had been thinking of $id as a number (with the possibility of 0 as a valid ID), rather than a string containing digits, which is what it is. Simply testing `if ($id)` is more appropriate. Will fix.

jeffbi replied:
This function has been removed

rmunn committed:
Fix Unicode in license_thirdparty_proprietary.txt (#2857)

This file contained many occurrences of U+F0A7, a character from the
designated Private Use Area, so its meaning is not defined and fonts
can assign any symbol to that character. Since the person who originally
wrote this file works at Microsoft, it seems reasonable to assume that
he was using the Wingdings font, which has a bullet character at U+F0A7.
Thus, replacing these characters with the proper Unicode bullet (U+2022)
seems the right way to fix this.

daxian-dbw committed:
Fix powershell class to use the current EngineSessionState for execution. (#2837)

When a PS class is defined in a module and the module gets reloaded, the class would still use the SessionState from the old module for execution, and thus it doesn't reflect changes to the module state during the reload. This fix is to make sure we always use the current EngineSessionState for PS class execution.

Commit:
More changes per code review

  Changes include:

    * Remove superfluous Windows version-checking code from cmdlets
    * Remove PowerShell functions borrowed from blog and re-implement with in-script C#
    * Add mandatory and validity checks to parameters to shared test functions
    * Modify comment about first record in exported files to provide additional information
    * Provide both "CI" and "Feature" tagged Describes in each cmdlets' test suite
    * Use Join-Path to create file paths
    * Use $testDrive for output file paths
    * Skip wild-card tests on non-English systems

Commit:
More changes per code review

  Changes include:

    * Remove superfluous Windows version-checking code from cmdlets
    * Remove PowerShell functions borrowed from blog and re-implement with in-script C#
    * Add mandatory and validity checks to parameters to shared test functions
    * Modify comment about first record in exported files to provide additional information
    * Provide both "CI" and "Feature" tagged Describes in each cmdlets' test suite
    * Use Join-Path to create file paths
    * Use $testDrive for output file paths
    * Skip wild-card tests on non-English systems

jeffbi commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/CounterTestHelperFunctions.ps1:
```diff
@@ -0,0 +1,310 @@
+#
+# This function came from an article in PowerShellMagazine by Tobias Weltner
+# http://www.powershellmagazine.com/2013/07/19/querying-performance-counters-from-powershell/
+Function Get-PerformanceCounterLocalName
```
This function has been removed

jeffbi commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/CounterTestHelperFunctions.ps1:
```diff
@@ -0,0 +1,310 @@
+        $counters = (Get-ItemProperty -Path $key -Name Counter).Counter
+        $all = $counters.Count
+
+        for ($i = 0; $i -lt $all; $i += 2)
```
This function has been removed

jeffbi commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/CounterTestHelperFunctions.ps1:
```diff
@@ -0,0 +1,310 @@
+# Exporting lops fractional milliseconds off the time stamp,
+# so simply comparing the DateTime values isn't sufficient
+function DateTimesAreEqualish($dtA, $dtB)
+{
```
Fixed

jeffbi commented on src/Microsoft.PowerShell.Commands.Diagnostics/CommonUtils.cs:
```diff
@@ -111,13 +145,39 @@ public static uint FormatMessageFromModule(uint lastError, string moduleName, ou
+
+            osVerEx.OSVersionInfoSize = Marshal.SizeOf<OSVERSIONINFOEX>(osVerEx);
+            if (!GetVersionEx(ref osVerEx))
+                throw new Win32Exception(Marshal.GetLastWin32Error());
```
This code has been removed

daxian-dbw committed:
Fix try/catch to choose the more specific exception handler (#2429)

* Refactor 'FindMatchingHandler' to get rid of recursion
Search state can be saved and used throughout the searching process without the recursion.

* Rank the exception types based on how specific they are

* Refactor code to find handler based on the rank of given handlers

Commit:
Merge remote-tracking branch 'refs/remotes/origin/perf-counters' into perf-counters

Francisco-Gamino commented:
Closing PR. This code change has been moved to https://github.com/PowerShell/PowerShell/pull/2868 to resolve merging conflicts, add logic to skip the tests on Linux and OSX, as well as to squash the commits to have one for product and one for test changes.
